notation/notation.go

```diff
@@ -0,0 +1,314 @@
+// remote.
+// The descriptor of the sign content is returned upon sucessful signing.
+func Sign(ctx context.Context, signer Signer, repo registry.Repository, opts SignOptions) (Descriptor, error) {
+	ociDesc, err := repo.Resolve(ctx, opts.ArtifactReference)
+	if err != nil {
+		return Descriptor{}, err
+	}
+	desc := notationDescriptorFromOCI(ociDesc)
+	sig, signerInfo, err := signer.Sign(ctx, desc, opts.SignatureMediaType, opts)
+	if err != nil {
+		return Descriptor{}, err
+	}
+	annotations, err := generateAnnotations(signerInfo)
+	if err != nil {
+		return Descriptor{}, err
+	}
+	_, _, err = repo.PushSignature(ctx, sig, opts.SignatureMediaType, ociDesc, annotations)
+	if err != nil {
+		return Descriptor{}, err
+	}
+
+	return desc, nil
+}
+
+// VerifyOptions contains parameters for Verifier.Verify.
+type VerifyOptions struct {
+	// ArtifactReference is the reference of the artifact that is been
+	// verified against to.
+	ArtifactReference string
+
+	// SignatureMediaType is the envelope type of the signature.
+	// Currently both `application/jose+json` and `application/cose` are
+	// supported.
+	SignatureMediaType string
+
+	// PluginConfig is a map of plugin configs.
+	PluginConfig map[string]string
+
+	// VerificationLevel encapsulates the signature verification preset and its
+	// actions for each verification type
+	VerificationLevel *trustpolicy.VerificationLevel
+}
+
+// ValidationResult encapsulates the verification result (passed or failed)
+// for a verification type, including the desired verification action as
+//
+//	specified in the trust policy
+type ValidationResult struct {
+	// Success is set to true if the verification was successful
+	Success bool
+	// Type of verification that is performed
+	Type trustpolicy.ValidationType
+	// Action is the intended action for the given verification type as defined
+	// in the trust policy
+	Action trustpolicy.ValidationAction
+	// Err is set if there are any errors during the verification process
+	Error error
+}
+
+// VerificationOutcome encapsulates a signature blob's descriptor, its content,
+// the verification level and results for each verification type that was
+// performed.
+type VerificationOutcome struct {
+	SignatureBlobDescriptor *ocispec.Descriptor
+	// EnvelopeContent contains the details of the digital signature and
+	// associated metadata
+	EnvelopeContent *signature.EnvelopeContent
+	// VerificationLevel describes what verification level was used for
+	// performing signature verification
+	VerificationLevel *trustpolicy.VerificationLevel
+	// VerificationResults contains the verifications performed on the signature
+	// and their results
+	VerificationResults []*ValidationResult
+	// SignedAnnotations contains arbitrary metadata relating to the target
+	// artifact that was signed
+	SignedAnnotations map[string]string
+	// Error that caused the verification to fail (if it fails)
+	Error error
+}
```

Comment:
Those structures should be in the `verifier` package.

Reply:
Both notation and verifier package are using these structures. verifier package also imports notation package. Putting these structures in the verifier package would cause import cycle.

Reply:
We can discuss offline if there are technical issue like resolving an import cycle.

Reply:
> BTW, what's reason would `verifier` package import package `notation`?

Oh, I see `verifier.New()` returns a `notation.Verifier`.